File: c++/ImplementStackUsingQueues.cpp
#include <iostream>
#include <queue>

class MyStack {
private:
	std::queue<int> queue;
public:
    /** Initialize your data structure here. */
    MyStack() {
        
    }
    
    /** Push element x onto stack. */
    void push(int x) {
        queue.push(x);
    }
    
    /** Removes the element on top of the stack and returns that element. */
    int pop() {
        for (int i = 0; i < queue.size() - 1; i++) {
        	queue.push(queue.front());
        	queue.pop();
        }
        int result = queue.front();
        queue.pop();
        return result;
    }
    
    /** Get the top element. */
    int top() {
        return queue.back();
    }
    
    /** Returns whether the stack is empty. */
    bool empty() {
        return queue.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * bool param_4 = obj.empty();
 */

int main() {
	MyStack obj;
	obj.push(1);
	obj.push(2);
	obj.push(3);
	obj.push(4);
	int param_2 = obj.pop();
    std::cout << "The element from in front of stack removed is: " << param_2 << std::endl;
    int param_3 = obj.top();
    std::cout << "The front element is: " << param_3 << std::endl;
    bool param_4 =  obj.empty();
    if (param_4 == true) {
        std::cout << "The stack is empty." << std::endl;
    } else {
        std::cout << "The stack is not empty." << std::endl;
    }
    return 0;
}

File: javascript/implement-stack-using-queues.js
class Node {
    constructor(val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

var MyStack = function() {
    this.head = null;
    this.length = 0;
};

/** 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function(x) {
    let node = new Node(x);
    if (this.empty()) {
        this.head = node;
    } else {
        this.head.prev = node;
        node.next = this.head;
        this.head = node;
    }
    this.length++;
};

/**
 * @return {number}
 */
MyStack.prototype.pop = function() {
    if (this.empty()) {
        return null;
    }
    let val = this.head.val;
    if (this.length === 1) {
        this.head = null;
    } else {
        this.head.next.prev = null;
        this.head = this.head.next;
    }
    this.length--;
    return val;
};

/**
 * @return {number}
 */
MyStack.prototype.top = function() {
    if (this.empty()) {
        return null;
    }
    return this.head.val;
};

/**
 * @return {boolean}
 */
MyStack.prototype.empty = function() {
    return !this.length;
};

/** 
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */